Refactor make_fx to better support hop subgraph tracing

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125363
* __->__ #125267

Code movement + minor rewrites. We extract the states of make_fx out and encapsulate them into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs, the actual logic for tracing subgraph is in the next diff.

Test Plan:
Existing tests.

ydwu4 committed:
Refactor make_fx to better support hop subgraph tracing

[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"




[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"




[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"




[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"




[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


This is to replace https://github.com/pytorch/pytorch/pull/122972.

We extract the states of make_fx out into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs. The entry point for tracings subgraphas is 'reenter_make_fx(fn)'.

Test Plan:
Existing tests and remove a related task on torch function.



[ghstack-poisoned]

ydwu4 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1143,91 +1143,96 @@ def create_node(self, *args, **kwargs):
 
-    @functools.wraps(f)
-    def wrapped(*args):
+    # TODO: change to a context mananer _override_defaults_based_on_inputs(f, args)
```
remove comment

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


Code movement + minor rewrites. We extract the states of make_fx out into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs. The entry point for tracings subgraphas is 'reenter_make_fx(fn)'.

Test Plan:
Existing tests and remove a related task on torch function.



[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


Code movement + minor rewrites. We extract the states of make_fx out into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs. The entry point for tracings subgraphas is 'reenter_make_fx(fn)'.

Test Plan:
Existing tests and remove a related task on torch function.



[ghstack-poisoned]

ydwu4 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1143,146 +1143,214 @@ def create_node(self, *args, **kwargs):
+
+        # All context managers and their states should be initialized before tracing based on the inputs
+        # and configurations.
+        # After tracing, thier states should be cleaned except for shape_env.
```
shape_env is leaked out because of `t.shape_env = self.fake_tensor_mode.shape_env`

Chillee reviewed (commented):
Can we get some tests of this? Or are you relying purely on the existing test suite?

ydwu4 commented:
yeah, this pr is supposed to be a code movement only change. Existing tests should suffice. Any suspicious places you want to test?

Chillee commented:
Ok looks fine to me.

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


Code movement + minor rewrites. We extract the states of make_fx out and encapsulate them into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs, the actual logic for tracing subgraph is in the next diff.

Test Plan:
Existing tests.



[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


Code movement + minor rewrites. We extract the states of make_fx out and encapsulate them into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs, the actual logic for tracing subgraph is in the next diff.

Test Plan:
Existing tests.



[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


Code movement + minor rewrites. We extract the states of make_fx out and encapsulate them into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs, the actual logic for tracing subgraph is in the next diff.

Test Plan:
Existing tests.



[ghstack-poisoned]

ydwu4 committed:
Update on "Refactor make_fx to better support hop subgraph tracing"


Code movement + minor rewrites. We extract the states of make_fx out and encapsulate them into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs, the actual logic for tracing subgraph is in the next diff.

Test Plan:
Existing tests.



[ghstack-poisoned]
